lib/options/showOption: fix quoting of attr-names that are not identifiers

Ma27 opened the pull request:
###### Description of changes
Personally, I think that warnings such as

    warning: The option `services.redis.enable' defined in `/home/ma27/Projects/nixpkgs/test.nix@node-vm' has been renamed to `services.redis.servers..enable'.

are fairly confusing because of the `..` and it's more correct to actually quote that. With this change the warning now looks like this:

    warning: The option `services.redis.enable' defined in `/home/ma27/Projects/nixpkgs/test.nix@node-vm' has been renamed to `services.redis.servers."".enable'.

While implementing that I realized that you'd have a similar problem whenever you use attribute-names that aren't identifiers, e.g.

    services.nginx.virtualHosts."example.org".locations."/".invalid = 23;

now results in the following error:

    error: The option `interactive.nodes.vm.services.nginx.virtualHosts."example.org".locations."/".invalid' does not exist. Definition values:
           - In `/home/ma27/Projects/nixpkgs/test.nix@node-vm': 23

Of course there are some corner-cases where this won't work: when generating the manual, you display submodules like this:

    services.nginx.virtualHosts.<name>

Since `<name>` isn't a value, but an indicator for a submodule, it must not be quoted. This also applies to the following identifiers:

* `*` for `listOf submodule`
* `<function body>` for `functionTo`

This might not be correct if you actually have a submodule with an attribute name called `<name>`, but I think it's an improvement over the current situation and for this you'd probably need to make even more complex changes to the module system.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Closes https://github.com/NixOS/nixpkgs/issues/152649

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth commented:
We shouldn't have been mixing placeholders and unescaped identifiers in the first place.
Instead, we could represent paths like `[ "etc" { placeholder = "<name>"; } "mode" ]`. This makes the code accurate and moves type-specific logic back to the types where it belongs.

Ma27 commented:
@roberth do you think it's possible to implement that without noticeable breaking changes? If yes, I'm happy to give it a try, but otherwise I'd say that this is the pragmatic solution to improve the situation a little bit at least :)

roberth commented on lib/options.nix:
```diff
@@ -323,7 +323,13 @@ rec {
         escaped = lib.strings.escapeNixString part;
-      in if escaped == "\"${part}\""
+        isIdentifier = x: builtins.match "^[a-zA-z_][a-zA-Z0-9_'-]*$" x != null;
+        specialIdentifiers = [
```
```suggestion
        # We assume that these are "special values" and not real configuration data.
        # If it is real configuration data, it is rendered incorrectly.
        specialIdentifiers = [
```

roberth reviewed (commented):
While I do think the correct solution is feasible, I don't mind getting there in steps.

Ma27 committed:
lib/types: always use `<function body>` instead of `[function body]` to indicate a function inside an option structure

The motivation is to have a single identifier for that. Useful for the
next commit where I'll try to escape option-parts correctly (options can
be any kind of strings, but unless these are Nix identifiers, they must
be quoted).

Since `<function body>` (or `<name>`/`*`) are special identifiers in
error messages and the manual, we need a unique way to mark an option
part as function call because these are not to be quoted.

Ma27 committed:
lib/options/showOption: fix quoting of attr-names that are not identifiers

Personally, I think that warnings such as

    warning: The option `services.redis.enable' defined in `/home/ma27/Projects/nixpkgs/test.nix@node-vm' has been renamed to `services.redis.servers..enable'.

are fairly confusing because of the `..` and it's more correct to
actually quote that. With this change the warning now looks like this:

    warning: The option `services.redis.enable' defined in `/home/ma27/Projects/nixpkgs/test.nix@node-vm' has been renamed to `services.redis.servers."".enable'.

While implementing that I realized that you'd have
a similar problem whenever you use attribute-names that aren't
identifiers, e.g.

    services.nginx.virtualHosts."example.org".locations."/".invalid = 23;

now results in the following error:

    error: The option `interactive.nodes.vm.services.nginx.virtualHosts."example.org".locations."/".invalid' does not exist. Definition values:
           - In `/home/ma27/Projects/nixpkgs/test.nix@node-vm': 23

Of course there are some corner-cases where this won't work: when
generating the manual, you display submodules like this:

    services.nginx.virtualHosts.<name>

Since `<name>` isn't a value, but an indicator for a submodule, it must
not be quoted. This also applies to the following identifiers:

* `*` for `listOf submodule`
* `<function body>` for `functionTo`

This might not be correct if you actually have a submodule with an
attribute name called `<name>`, but I think it's an improvement over the
current situation and for this you'd probably need to make even more
complex changes to the module system.

Ma27 commented on lib/options.nix:
```diff
@@ -323,7 +323,13 @@ rec {
       let
         escaped = lib.strings.escapeNixString part;
-      in if escaped == "\"${part}\""
+        isIdentifier = x: builtins.match "^[a-zA-z_][a-zA-Z0-9_'-]*$" x != null;
```
Admittedly I didn't bother to look and checked rnix-parser to see if I was missing some details about what identifiers can contain.

But you're right, I updated the code accordingly.

Ma27 commented on lib/types.nix:
```diff
@@ -616,8 +616,8 @@ rec {
       merge = loc: defs:
-        fnArgs: (mergeDefinitions (loc ++ [ "[function body]" ]) elemType (map (fn: { inherit (fn) file; value = fn.value fnArgs; }) defs)).mergedValue;
-      getSubOptions = prefix: elemType.getSubOptions (prefix ++ [ "[function body]" ]);
+        fnArgs: (mergeDefinitions (loc ++ [ "<function body>" ]) elemType (map (fn: { inherit (fn) file; value = fn.value fnArgs; }) defs)).mergedValue;
+      getSubOptions = prefix: elemType.getSubOptions (prefix ++ [ "<function body>" ]);
```
Done.

Ma27 commented:
Thanks a lot for your review! I addressed all of your coments :)

roberth reviewed (approved):
Looks ok to me
@infinisil please review

infinisil reviewed (dismissed):
This currently breaks the manual build, since the XSLT processing relies on this in some weird way, you can check with

```
nix-build nixos/release.nix -A manualHTML.x86_64-linux
```

I previously broke the manual in almost the same way when I tried to change the implementation of `showOption` a bit in https://github.com/NixOS/nixpkgs/pull/82461, which was then subsequently reverted in https://github.com/NixOS/nixpkgs/pull/85241. There's also a bug report regarding `showOption` in https://github.com/NixOS/nixpkgs/issues/152649

Ma27 committed:
nixos/doc: fix build

Now we even have options like
`services.listmonk.database.settings."app.notify_emails"` shown
correctly (i.e. with quotes).

Ma27 commented:
@infinisil I replaced the quotes within the xslt now with a `_` and it appears to work fine. Options like `services.listmonk.database.settings."app.notify_emails"` are also shown correctly quoted in the manual now :)

roberth commented on nixos/lib/make-options-doc/options-to-docbook.xsl:
```diff
@@ -39,8 +39,8 @@
-                '*&lt; >[]:',
-                '_______'
+                '*&lt; >[]:&quot;',
+                '________'
```
This changes the identifiers, breaking inbound links. Could you remove the quotes instead of replacing them with underscores?

Ma27 replied:
Shouldn't that throw an error if `<xref linkend="" />` points to an unknown location?

roberth replied:
Then that would have been a problem before.
You'd have to define `options."" = mkOption ...` at the top level, which I don't think is something we have to worry about.

Ma27 replied:
> Then that would have been a problem before.

Yes. But considering that I can build the manual locally, doesn't that mean that there are no breaking links?

> You'd have to define options."" = mkOption ... at the top level, which I don't think is something we have to worry about.

Even though it's just a theoretical problem, having e.g. `settings."foo.bar"` and `settings.foo.bar` defined would cause a conflict when removing the quotes, that's why I kept them.

Or am I misunderstanding your point? :)

Ma27 replied:
ping @roberth

roberth replied:
I was more concerned about links from the web to the manual. Passing the internal validation will not detect such broken links.
It does appear that the impact is quite limited:

(`{+_+}` is git word diff syntax for "add underscore")

 - 138 ids relating to sr.ht such as `opt-services.sourcehut.settings.{+_+}sr.ht{+_+}.owner-name`
 - `opt-services.listmonk.database.settings.{+_+}app.notify_emails{+_+}`
 - `opt-services.listmonk.database.settings.{+_+}bounce.mailboxes{+_+}`
 - `opt-services.listmonk.database.settings.{+_+}privacy.domain_blocklist{+_+}`
 - `opt-services.listmonk.database.settings.{+_+}privacy.exportable{+_+}`
 - `opt-services.xserver.windowManager.{+_+}2bwm{+_+}.enable`

This seems like a small price to pay for a simpler and more injective id generating function.

Ma27 commented:
@roberth so is it fine to keep that as-is? And: anything else tbd here? :)
